llvm/lib/Target/AMDGPU/SILoadStoreOptimizer.cpp

```diff
@@ -1363,6 +1363,15 @@ SILoadStoreOptimizer::checkAndPrepareMerge(CombineInfo &CI,
   return Where;
 }
 
+static bool isCompatibleAlignSubReg(const TargetRegisterClass *RC, unsigned Idx,
+                                    const GCNSubtarget *STM,
+                                    const SIRegisterInfo *TRI) {
+  if (!STM->needsAlignedVGPRs() || !TRI->isVGPRClass(RC) ||
+      !TRI->isProperlyAlignedRC(*RC) || AMDGPU::getRegBitWidth(*RC) == 32)
+    return true;
+  return !(TRI->getChannelFromSubReg(Idx) & 0x1);
```

Comment:
You shouldn't need to do any of this special case hardcoding. Just check if the register class of the virtual register or operand is compatible with the new class

Reply:
You should use none of the functions here; not needsAlignedVGPRs, not isProperlyAlignedRC, and probably not getChannelFromSubReg or getRegBitWidth either. 

You either want RC->hasSubClassEq() or TRI->getCommonSubClass depending on the context. 

In this case it looks like getTargetRegisterClass isn't returning the correct class to use

Reply:
Removed conditional altogether